### ISSUE TITLE
Special Characters in Filenames - Filename fallback must only contain ASCII characters

### ISSUE THREAD

**Issue by mvscheidt** [contao/core-bundle#732]
It seems there is a problem downloading files with Special Characters (I noticed that with Umlauts).
Direct Links to the files do not suffer from this - only the download Links!

This happens both in the Frontend (Downloads - it's where a customer noticed it) aswell as the FileManager in the Backend.

We noticed this Problem in a System with Core-Bundle 4.3.3 (PHP 7.1.1) and in another System where Core-Bundle 4.3.1 (PHP 7.0.16-1~dotdeb+8.1) is installed.

Things to note:

I have not tested this in 4.3.6 - I wanted to test it in the Contao Demo (assuming it is up to date) but file uploads are deactivated there (for understandable reasons).

I'm not entirely sure if this may be an incompatibility between Symfony <-> Contao or just a Contao Bug because If memory serves me right, Contao used to "rename" files (at least in one System I just checked on 3.2.16 it does this - so "Über.txt" turns into "Ueber.txt") but that doesn't happen for the above two Contao 4 versions. Not sure if that is a bug or it was changed.

Example Files I used:

[Über.txt](https://github.com/contao/core/files/864774/Uber.txt)
[Ueber.txt](https://github.com/contao/core/files/864775/Ueber.txt)

Upload them via the FileManager.

* Try to Download "Ueber.txt" from the Filemanager Info Overlay - works as expected and Download is created
* Try to Download "Über.txt" from the Filemanager Info Overlay - no download is created, and the overlay displays an error.

Error output in Logfile

```
[2017-03-23 14:02:56] app.CRITICAL: An exception occurred. {"exception":"[object] (InvalidArgumentException(code: 0): The filename fallback must only contain ASCII characters. at vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php:251)"} []
```

----

Screenshots from the Backend.

![file_manager_-_contao_open_source_cms_4 3_-_2017-03-23_14 02 13](https://cloud.githubusercontent.com/assets/6872430/24249050/1746637e-0fd2-11e7-8524-1a91bc8bcc88.png)

Screenshots from the Overlay __After__ clicking on "Download the file":

![screenshot from 2017-03-23 14-02-46](https://cloud.githubusercontent.com/assets/6872430/24249365/3f8755ae-0fd3-11e7-84bd-239c55efda53.png)
![file_manager_-_contao_open_source_cms_4 3_-_2017-03-23_14 03 20](https://cloud.githubusercontent.com/assets/6872430/24249364/3f8671b6-0fd3-11e7-925c-96960e1a9baf.png)

**Comment by leofeyer** [comment]
I cannot reproduce the issue on my system. Both files can be downloaded without a problem.

**Comment by asaage** [comment]
might be related to https://github.com/contao/core/issues/8295
Check if it happens if the first letter is not a special-character!
Check the tl_file record in the database!

**Comment by mvscheidt** [comment]
As It happens a colleague is currently installing 4.3.6 - I will try it once it's set up. 

> Check if it happens if the first letter is not a special-character!

It still happens. Tried with "Uüber.txt"

> Check the tl_file record in the database!

It's also in the database.

@DanielSchwiperich just send me this aswell: http://stackoverflow.com/questions/20094286/laravel-4-the-filename-fallback-must-only-contain-ascii-characters

**Comment by DanielSchwiperich** [comment]
> Symfony HttpFoundation, which manages file downloads, requires the file being downloaded to have an ASCII file name.

That is the interesting part from the link. 
We'll have a look at the implementation after testing 4.3.6 and if the error still occurs

**Comment by mvscheidt** [comment]
4.3.6 throws the same Exception as I described in the issue description for me.

Will have to check what's going on here.

**Comment by DanielSchwiperich** [comment]
@leofeyer Just to be 100% sure, you tested the download functionality which is using the binaryResponse and not the direct file link (that for works us)

The exception thrown is pretty clear that symfony expects an ascii name and if nots convert as such it should not work on any system. 

```
[2017-03-23 14:02:56] app.CRITICAL: An exception occurred. {"exception":"[object] (InvalidArgumentException(code: 0): The filename fallback must only contain ASCII characters. at vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php:251)"} []
```

**Comment by DanielSchwiperich** [comment]
It looks to be related contao/core#8295

If you upload a file with an Umlaut within it gets saved with that Umlaut. That will cause downloads to break since it gives the basename as fallback name to the symfony BinaryResponse which only allows ascii chars. 

https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/File.php#L787

As soon as you rename the file the umlaute get replaced and the downloads then work. 
(Just saving doesn't do anything)
Calling the routine which processes the name on renaming also on upload would solve the problem I think (haven't look for that one yet)

**Comment by mvscheidt** [comment]
I just checked another customer or ours which still uses 3.5.12 - the File can be downloaded there. Compared to 3.2.16 the file is not renamed after Uploading anymore (so the Umlaut does not get removed anymore since at least 3.5.12).

Currently using Symfony 3.2.6 and Contao 4.3.7 on fpm-fcgi Environment with PHP 7.1.0 (Versions from the Debug Toolbar)

Just to make sure we are all really on the same page - the following type of download link causes the issue (for us at least)
* http://domain.tld/app_dev.php/contao/popup?src=ZmlsZXMvJUMzJTlDYmVyLnR4dA==&download=1

and this one (obviously) does not
* http://domain.tld/files/%C3%9Cber.txt

(Keep in mind we noticed this in the Frontend Downloads initially, I'm just testing around in the backend currently)

The latter one works as @DanielSchwiperich pointed out because its just not passing through the `BinaryFileResponse`, or to be more exact through the following method(s):

* https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/BinaryFileResponse.php#L178
* https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L288

Still it's weird that it works for @leofeyer , because the fallback should always be set when using the "Download the file" Button in the overlay, at least from a quick glance over the code in Contao / Symfony.

Reason being that the $filename param in Contao is only set if you pass a filename to the `sendToBrowser()` Method of Contao, which does not happen for the `BackendPopup` Class (it doesn't pass any name), see: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/controllers/BackendPopup.php#L96

If I remove the param I can actually download the file which also retains its original name including the Umlaut: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/File.php#L787

The following change in `File.php` also works for me and it would retain the possibility to rename the file when calling Contaos `sendToBrowser()` method (e.g. from Modules or somewhere else in the Code, though I have not tested that).

```
        //use the basename if no filename is passed
        if ('' == $filename) {
            $filename = $this->basename;
        }

        $response->setContentDisposition
        (
            ResponseHeaderBag::DISPOSITION_ATTACHMENT,
            $filename
        );
```

It still bothers me though that it works for some (like @leofeyer ) but not for us, so I wouldn't really want to be so quick and say this is one valid way to fix it. Especially without knowing why the fallback was configured in the first place.

I wonder if this may also be something related to some locale settings for php - could someone who does not have this problem tell me what kind of locale the php process uses for them when downloading the file?

**Comment by leofeyer** [comment]
Fixed in 95e8ac36ff2b8c585ea217a0ca0f3cb1cab505c6.

The reason I could not reproduce the issue was that Chrome supports the `download=""` attribute and thus the binary response was never triggered. (Even though @DanielSchwiperich gave me a heads up on this in his [comment](https://github.com/contao/core-bundle/issues/732#issuecomment-288757030)).
